src/inet/InetInterface.cpp

```diff
@@ -226,6 +226,125 @@ void CloseIOCTLSocket(void)
 
 #if WEAVE_SYSTEM_CONFIG_USE_SOCKETS
 
+#if __ANDROID_API__ >= 21 && __ANDROID_API__ < 24
+
+static struct if_nameindex * backport_if_nameindex(void);
+static void backport_if_freenameindex(struct if_nameindex *);
+
+static void backport_if_freenameindex(struct if_nameindex *inArray)
+{
+    if (inArray == NULL)
+        return;
+
+    for (size_t i = 0; inArray[i].if_index != 0; i++)
+    {
+        if (inArray[i].if_name != NULL)
+            free(inArray[i].if_name);
+    }
+
+    free(inArray);
+}
+
+static struct if_nameindex * backport_if_nameindex(void)
+{
+    int err;
+    unsigned index;
+    size_t intfIter = 0;
+    size_t maxIntfNum = 0;
+    size_t numIntf  = 0;
+    size_t numAddrs = 0;
+    struct if_nameindex *retval = NULL;
+    struct if_nameindex *tmpval = NULL;
+    struct ifaddrs * addrList = NULL;
+    struct ifaddrs * addrIter = NULL;
+    const char * lastIntfName = "";
+
+    err = getifaddrs(&addrList);
+    VerifyOrExit(err >= 0,);
+
+    // coalesce on consecutive interface names
+    for (addrIter = addrList; addrIter != NULL; addrIter = addrIter->ifa_next)
+    {
+        numAddrs++;
+        if (strcmp(addrIter->ifa_name, lastIntfName) == 0)
+            continue;
+        numIntf++;
+        lastIntfName = addrIter->ifa_name;
+    }
+
+    tmpval = (struct if_nameindex *)malloc((numIntf + 1) * sizeof(struct if_nameindex));
+    VerifyOrExit(tmpval != NULL,);
+    memset(tmpval, 0, (numIntf + 1) * sizeof(struct if_nameindex));
+
+    lastIntfName = "";
+    for (addrIter = addrList; addrIter != NULL; addrIter = addrIter->ifa_next)
+    {
+        if (strcmp(addrIter->ifa_name, lastIntfName) == 0)
+            continue;
+
+        index = if_nametoindex(addrIter->ifa_name);
+        if (index != 0)
+        {
+            tmpval[intfIter].if_index = index;
+            tmpval[intfIter].if_name  = strdup(addrIter->ifa_name);
+            intfIter++;
+        }
+        lastIntfName = addrIter->ifa_name;
+    }
+
+    // coalesce on interface index
+    maxIntfNum = 0;
+    for (size_t i = 0; tmpval[i].if_index != 0; i++)
```

Comment:
I may be mistaken but I think you could just use the value of `intfIter` here rather than looping through again? That value is only incremented when index != 0 so it should be equivalent.

Reply:
Not quite.  The list returned from `getifaddrs` contains an entry for each IP address in the system.  I've tried to make very few assumptions about the ordering of that list, and I've been trying to make no assumptions on whether the `if_index` is dense. 

1. the first pass through the list from `getifaddrs` coalesces the on consecutive entries with the same interface name.  Arguably this could be an optional step.  This (along with just the projection from `struct ifaddrs` to `struct if_nameindex` ) takes place in lines 265..293. Note that at this point in time, `tmpval` is dense, it may contain duplicate entries for the same interface, and there is no relationship between the number of entries and the max index of the interface. 
2. the second pass (lines 295..311) reshuffles the entries: we use the `if_index` to index into the array to be returned.   This creates a duplicate free array of interfaces, but this array is potentially sparse -- if there are gaps between interface indeces, the array will contain 0s in those entries, and thus not conform to the expected output of `if_nameindex`. 
3. the third pass (lines 316..321)  coalesces the empty entries.   At the end of the pass, all the entries with `0` in `if_index` are at the end of the array.

Reply:
Got it, thanks for the explanation!